### PR TITLE
Fixed-size task: implement aggregation job creation.

### PR DESCRIPTION
**Pull request by branlwyd**
Part of #468.

### CODE REVIEW

File: aggregator/src/aggregator/aggregation_job_creator.rs [hunk 1]
```diff
@@ -347,6 +380,164 @@ impl<C: Clock + 'static> AggregationJobCreator<C> {
...
+        max_batch_size: usize,
+    ) -> anyhow::Result<()>
+    where
+        for<'a> &'a A::AggregateShare: Into<Vec<u8>>,
+        A::PrepareMessage: Send + Sync,
+        A::PrepareState: Send + Sync + Encode,
+        A::OutputShare: Send + Sync,
+        for<'a> &'a A::OutputShare: Into<Vec<u8>>,
+    {
+        Ok(self
```
Comment (tgeoghegan) [inline aggregator/src/aggregator/aggregation_job_creator.rs:399]: nit: if you did
```
let ret = self.datastore.run_tx(...);
Ok(ret)
```
...then you could de-indent the bulk of the method by one tab, which maybe gives `cargo fmt` a little more breathing room to lay out the code, but it's no big deal.

File: aggregator/src/aggregator/aggregation_job_creator.rs [hunk 2]
```diff
@@ -347,6 +380,164 @@ impl<C: Clock + 'static> AggregationJobCreator<C> {
...
+                        //    aggregating more than max_batch_size reports together.
+                        // Choose the maximal size meeting all of these requirements.
+                        let aggregation_job_size = [
+                            unaggregated_report_ids.len(),
+                            this.max_aggregation_job_size,
+                            max_batch_size - max_size,
+                        ]
+                        .into_iter()
+                        .min()
+                        .unwrap();
```
Comment (tgeoghegan) [inline aggregator/src/aggregator/aggregation_job_creator.rs:445]: It'd be cool if there was some kind of `InfiniteIterator` trait for iterators that are guaranteed to always yield a value so that one could iterate without all the `unwrap()`s... but that's not a review comment on this change!

File: aggregator/src/aggregator/aggregation_job_creator.rs [hunk 3]
```diff
@@ -347,6 +380,164 @@ impl<C: Clock + 'static> AggregationJobCreator<C> {
...
+
+                        // Generate an aggregation job, then update batch metadata & continue.
+                        let aggregation_job_id = random();
+                        debug!(
+                            task_id = %task.id(),
+                            batch_id = %batch.id(),
+                            %aggregation_job_id,
+                            report_count = aggregation_job_size,
+                            "Creating aggregation job"
+                        );
```
Comment (tgeoghegan) [inline aggregator/src/aggregator/aggregation_job_creator.rs:482]: Have we plumbed a metrics collector into this struct? If so, we should bump a counter for aggregation job creation. If not, let's put a TODO comment and file an issue about complementing `debug!` and `info!` prints with metrics.

File: aggregator/src/datastore.rs [hunk 4]
```diff
@@ -3619,6 +3726,56 @@ pub mod models {
     {
     }
 
+    /// An outstanding batch, which is a batch which has not yet started collection.
```
Comment (tgeoghegan) [inline aggregator/src/datastore.rs:3729]: Besides not yet having been collected, an `OutstandingBatch` can also still have reports assigned to it, right?

File: aggregator/src/aggregator/aggregation_job_creator.rs [hunk 5]
```diff
@@ -347,6 +380,164 @@ impl<C: Clock + 'static> AggregationJobCreator<C> {
...
+                        }));
+
+                    // Main loop: repeatedly consume some of the unaggregated report IDs to generate
+                    // an aggregation job, assigning it to an existing batch which has need of
+                    // reports, or a new batch if no existing batch needs reports.
+                    let mut aggregation_jobs = Vec::<AggregationJob<L, FixedSize, A>>::new();
+                    let mut report_aggregations = Vec::<ReportAggregation<L, A>>::new();
+                    let mut new_batches = Vec::new();
+                    let (mut is_batch_new, mut batch) = batch_iter.next().unwrap();
+                    let mut max_size = batch.max_size();
```
Comment (tgeoghegan) [inline aggregator/src/aggregator/aggregation_job_creator.rs:428]: It's a bit confusing resolving the notion of an `OustandingBatch`'s `max_size` against the task's `max_size`. IIUC `OutstandingBatch.max_size` is the biggest batch that could be constructed based on the reports currently assigned to the batch, but in fact the batch could still get bigger if we end up assigning more reports, right? Maybe `OutstandingBatch.potential_size` or (perhaps too wordy) `OutstandingBatch.current_potential_size` is more clear?

**Review by tgeoghegan — CHANGES_REQUESTED** [review]
Looks great, just a handful of suggestions that I don't think are critical.

**Review by divergentdave — COMMENTED** [review]
Thinking ahead to database performance implications, I think `read_batch_size()`will take out a lot of predicate locks on `report_aggregations` as it touches random pages during index scans. We don't need to address this now, as this is part of a larger known issue. We may want to denormalize the number of report aggregations in different states in an aggregation job into the `aggregation_jobs` table, to trade off a few writes for many more reads.

File: aggregator/src/datastore.rs [hunk 6]
```diff
@@ -2342,6 +2344,111 @@ ORDER BY id DESC
...
+            .prepare_cached(
+                "WITH batch_report_aggregation_statuses AS
+                    (SELECT report_aggregations.state, COUNT(*) AS count FROM report_aggregations
+                     JOIN aggregation_jobs ON report_aggregations.aggregation_job_id = aggregation_jobs.id
+                     WHERE aggregation_jobs.task_id = (SELECT id FROM tasks WHERE task_id = $1)
+                     AND aggregation_jobs.batch_identifier = $2
+                     GROUP BY report_aggregations.state)
+                SELECT
+                    (SELECT SUM(count)::BIGINT FROM batch_report_aggregation_statuses
+                     WHERE state IN ('FINISHED')) AS min_size,
```
[the remark this answers is not in the document]
Comment (branlwyd) [reply aggregator/src/datastore.rs:2426]: Not really -- I was just matching structure with the following few lines. (I would be pretty surprised if the query plan was any different than using `==`.)

File: aggregator/src/aggregator/aggregation_job_creator.rs [hunk 7]
```diff
@@ -23,11 +26,12 @@ use prio::{
     },
 };
 use rand::{random, thread_rng, Rng};
-use std::{collections::HashMap, convert::Infallible, sync::Arc, time::Duration};
+use std::{collections::HashMap, convert::Infallible, iter, sync::Arc, time::Duration};
 use tokio::{
     select,
     sync::oneshot::{self, Receiver, Sender},
     time::{self, Instant, MissedTickBehavior},
+    try_join,
```
[the remark this answers is not in the document]
Comment (branlwyd) [reply aggregator/src/aggregator/aggregation_job_creator.rs:34]: I can't think of a reason doing so would cause problems, but might as well be consistent -- I mostly-arbitrarily settled on using `tokio::try_join` throughout.

File: aggregator/src/aggregator/aggregation_job_creator.rs (hunk 1 above)
Comment (branlwyd) [reply aggregator/src/aggregator/aggregation_job_creator.rs:399]: I tried this out, but unfortunately `rustfmt` didn't actually de-indent anything -- all the lines after the edited line are indented equivalently.

(Since the `Ok` result will always be `()`, I tried returning `Ok(())` instead of assigning to `ret`, but unfortunately the results were the same.)

File: aggregator/src/aggregator/aggregation_job_creator.rs [hunk 8]
```diff
@@ -347,6 +380,164 @@ impl<C: Clock + 'static> AggregationJobCreator<C> {
...
+        A::PrepareState: Send + Sync + Encode,
+        A::OutputShare: Send + Sync,
+        for<'a> &'a A::OutputShare: Into<Vec<u8>>,
+    {
+        Ok(self
+            .datastore
+            .run_tx(|tx| {
+                let (this, task) = (Arc::clone(&self), Arc::clone(&task));
+                Box::pin(async move {
+                    let min_batch_size = usize::try_from(task.min_batch_size())?;
```
[the remark this answers is not in the document]
Comment (branlwyd) [reply aggregator/src/aggregator/aggregation_job_creator.rs:404]: Sure, makes sense. I also pedantically moved the conversions out of the transaction, on the idea that if there is a conversion problem, we're better served noticing it before we open a transaction than after. (pedantic because, on 64-bit archs, I'm pretty sure `u64` to `usize` conversion is infallible)

File: aggregator/src/aggregator/aggregation_job_creator.rs [hunk 9]
```diff
@@ -347,6 +380,164 @@ impl<C: Clock + 'static> AggregationJobCreator<C> {
...
+                            (true, OutstandingBatch::new(*task.id(), random(), 0, 0))
+                        }));
+
+                    // Main loop: repeatedly consume some of the unaggregated report IDs to generate
+                    // an aggregation job, assigning it to an existing batch which has need of
+                    // reports, or a new batch if no existing batch needs reports.
+                    let mut aggregation_jobs = Vec::<AggregationJob<L, FixedSize, A>>::new();
+                    let mut report_aggregations = Vec::<ReportAggregation<L, A>>::new();
+                    let mut new_batches = Vec::new();
+                    let (mut is_batch_new, mut batch) = batch_iter.next().unwrap();
```
[the remark this answers is not in the document]
Comment (branlwyd) [reply aggregator/src/aggregator/aggregation_job_creator.rs:427]: Added comments to both this & the other `unwrap` you mentioned. (I'm in favor of commenting all `unwrap`s as to why they are safe, in the cases they can't be avoided entirely.)

File: aggregator/src/aggregator/aggregation_job_creator.rs (hunk 2 above)
Comment (branlwyd) [reply aggregator/src/aggregator/aggregation_job_creator.rs:445]: 👍🏻

File: aggregator/src/aggregator/aggregation_job_creator.rs (hunk 3 above)
Comment (branlwyd) [reply aggregator/src/aggregator/aggregation_job_creator.rs:482]: Filed https://github.com/divviup/janus/issues/680.

File: aggregator/src/datastore.rs (hunk 4 above)
Comment (branlwyd) [reply aggregator/src/datastore.rs:3729]: Correct. It also only applies to `fixed-size` tasks. I noted both of these in the doc comment.

File: aggregator/src/aggregator/aggregation_job_creator.rs (hunk 5 above)
Comment (branlwyd) [reply aggregator/src/aggregator/aggregation_job_creator.rs:428]: Correct: `min_size` & `max_size` track the potential size of the batch, depending on how currently-incomplete aggregations resolve. Specifically, `min_size` tracks how many reports are included in the batch if all reports currently undergoing aggregation eventually fail (i.e. it is the number of report aggregations in state `FINISHED`). `max_size` tracks how many reports are included in the batch if all reports currently undergoing aggregation eventually succeed (i.e. it is the number of report aggregations in any non-terminal-failure state).

Let's chat about the names:
* I like not including `current` -- the reasoning here is that every model is the "current" state of the underlying data [at some point in time, I suppose], so that word is basically implied for every field, or at least every field that is effectively mutable over time. That is, by the same logic we would justify calling the `state` field contained by many data models `current_state` instead.
* `potential_size` has a couple of issues that I think need to be resolved to get to a good naming scheme:
  * An fair (IMO) interpretation is that every size between `min_size` and `max_size` (inclusive) is a "potential" size of the batch. We could call this `max_potential_size` but then we're kind of back to the original problem, I guess.
  * If we interpret the max size as the "potential" size, we still need a good name for the minimum size.
  * I considered just making a single field `potential_size: Range<usize>`, tracking the entire range of possible batch sizes. Unfortunately, Rust's `Range` type is exclusive, which would make the logic using this field more complex/error-prone (i.e. off-by-one errors) than otherwise necessary. Still, I can go ahead w/ this strategy if you think it wise.

IMO, `min_size` & `max_size` are OK since they are attached to an `OutstandingBatch` object. I did go ahead and update the variable naming in `aggregation_job_creator.rs` to attempt to ensure clarity between a task's max_batch_size & a batch's size.

Comment (divergentdave) [reply aggregator/src/aggregator/aggregation_job_creator.rs:428]: There is std::ops::RangeInclusive, FWIW

File: aggregator/src/aggregator/aggregation_job_creator.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (branlwyd) [reply aggregator/src/aggregator/aggregation_job_creator.rs:404]: Agh, unfortunately this makes the borrow-checker unhappy(!). Specifically, in the match arms, `max_batch_size` ends up being a borrow from `task`. Then trying to pass both `task` & `max_batch_size` (or `*max_batch_size`) causes the compiler to complain about task being used while borrowed. I think I could fix this by passing a `&Task` rather than an `Arc<Task>`, but the whole point of using `Arc` is that the `create_aggregation_jobs_for_fixed_size_task_no_param` needs one -- I would ultimately have to clone the `Arc`. I'd like to avoid incurring even likely-negligible runtime cost to avoid one line, so I moved the max_batch_size copy back out to `create_aggregation_jobs_for_task`.

File: aggregator/src/datastore.rs [hunk 10]
```diff
@@ -3619,6 +3726,56 @@ pub mod models {
     {
     }
 
+    /// An outstanding batch, which is a batch which has not yet started collection.
+    #[derive(Clone, Debug, PartialEq, Eq)]
+    pub struct OutstandingBatch {
+        /// The task ID for this outstanding task.
+        task_id: TaskId,
+        /// The batch ID for this outstanding task.
```
[the remark this answers is not in the document]
Comment (branlwyd) [reply aggregator/src/datastore.rs:3734]: 👍🏻 thanks, nice spot.

File: aggregator/src/datastore.rs [hunk 11]
```diff
@@ -3619,6 +3726,56 @@ pub mod models {
...
+            &self.batch_id
+        }
+
+        /// Gets the minimum possible size of this batch with current aggregation jobs (i.e. the
+        /// count of reports which have successfully completed the aggregation process).
+        pub fn min_size(&self) -> usize {
+            self.min_size
+        }
+
+        /// Gets the maximum possible size of this batch with current aggregatino jobs (i.e. the
```
[the remark this answers is not in the document]
Comment (branlwyd) [reply aggregator/src/datastore.rs:3771]: 👍🏻

File: aggregator/src/datastore.rs [hunk 12]
```diff
@@ -7089,6 +7219,154 @@ mod tests {
...
+        let (ds, _db_handle) = ephemeral_datastore(clock.clone()).await;
+
+        let (task_id, batch_id) = ds
+            .run_tx(|tx| {
+                let clock = clock.clone();
+                Box::pin(async move {
+                    let task = TaskBuilder::new(
+                        QueryType::FixedSize { max_batch_size: 10 },
+                        VdafInstance::Fake,
+                        Role::Helper,
```
[the remark this answers is not in the document]
Comment (branlwyd) [reply aggregator/src/datastore.rs:7236]: 👍🏻

**Comment by branlwyd** [comment]
> Thinking ahead to database performance implications, I think `read_batch_size()`will take out a lot of predicate locks on `report_aggregations` as it touches random pages during index scans. We don't need to address this now, as this is part of a larger known issue. We may want to denormalize the number of report aggregations in different states in an aggregation job into the `aggregation_jobs` table, to trade off a few writes for many more reads.

Agreed, though I think the fact that updates to aggregation job/report aggregations in the leader happen as "pure writes" (i.e. we have a transaction just for writing back updates to the aggregation jobs when they are stepped, reading happens in a different transaction & the VDAF computations happen outside of a transaction) will help to mitigate this -- the necessary retries will not be overly costly.

Another strategy would be to denormalize the counts into the `outstanding_batches` table (e.g. directly track `{min,max}_size`) -- this would lead to additional contention when an aggregation job is started or completes, but given the above argument about contention not being _too_ bad in this situation, we might be able to get away with it.

In any case, this is a lot of speculation on my part. :) Let's go with this approach for now and improve if we find we need to. This approach, at least, is less error-prone than trying to keep a counter updated at every point an aggregation job or its report aggregations can be updated.

File: aggregator/src/aggregator/aggregation_job_creator.rs (hunk 5 above)
Comment (branlwyd) [reply aggregator/src/aggregator/aggregation_job_creator.rs]: [I realized you're traveling, so I'm going to submit this PR without waiting to hear back on this issue, but poke this if you have an idea & I'll address in a followup PR.]

Comment (branlwyd) [reply aggregator/src/aggregator/aggregation_job_creator.rs]: I don't know how I missed that, I was looking specifically for it. Thanks, `OutstandingBatch` now has a single `size` method.

**Comment by branlwyd** [comment]
One more stray thought regarding performance: other than denormalization, another strategy would be to split `create_aggregation_jobs_for_fixed_size_task_no_param` into the following pieces:

* A read-only transaction which reads the required data (i.e. unaggregated report IDs, outstanding batches). [This would be another place justifying true read-only transaction support, avoiding any potential for restarts due to contention.]
* A computation of the outstanding batches/aggregation jobs/report aggregations to write, outside of a transaction.
* A write-only transaction which inserts the values computed in the previous step.

This would _sort of_ work out currently, since the current model is that there is a single aggregation job creator running. This model will likely need to evolve as we scale, but I imagine we could justify that there is a single aggregation job creator job responsible for generating aggregation jobs for a given task for quite a while.

I say it _sort of_ works out because, even though there should be only one aggregation job creator running at once, there is nothing enforcing this -- a simple operational misconfiguration causing multiple aggregation job creators to run at once (for a given task) could lead to some very strange behavior. We could mitigate this by implementing some change-detection in the write tx (bailing if it notices an unexpected write on rows it believes it is the only writer to); this may be simpler/more localized to implement than maintaining counters.

**Review by tgeoghegan — APPROVED** [review]
The changes LGTM.

re: database performance: I agree with moving forward with the current solution while acknowledging the potential for future changes. Could we capture the discussion of performance risks and solution strategies in an issue so it doesn't get lost when this PR is merged? It seems like we will eventually need to significantly overhaul the DB schema and datastore module, so it'd be nice to be able to quickly refer back to all our concerns and ideas when we tackle that work.
